projects/atomic-swap-service/headless-wallet.md

```diff
@@ -42,23 +34,37 @@ These identifier and password can then be shared with the other interested parti
 This integration with the swap service backend needs a way for the headless to store in memory the proposals that are currently being monitored by each of its initialized wallets.
 
 There are some considerations about this object:
-- An event listener will be created for each proposal to receive web-socket updates from the service
+- An event listener will be created for each wallet to receive web-socket updates from the service
 - More than one initialized wallet can be listening to the same proposal
-  - There should be no more than one websocket connection open for each proposal
-- Wallets that did not register/create a proposal should not receive updates about it
-
-To achieve this, a `listenedProposals` map will be created on the `services/wallets.service` module containing, in short, the proposal identifier, password, and a list of wallet-ids that are listening to it:
-```ts
-{
-	proposalId: string,
-	password: string,
-	wallets: string[],
-}
+- Wallets that did not register a proposal should not receive updates about it
+
+To achieve this, a `walletListenedProposals` map will be created on the `services/wallets.service` module.
+```js
+/**
+ * @typedef TxProposalConfig
+ * @property {string} id Proposal identifier on the Atomic Swap Service
+ * @property {string} password Password to access it
+ */
+
+/**
+ * A map of all the proposals for a wallet.
+ * The keys are the proposal ids
+ * @typedef {Record<string,TxProposalConfig>} WalletListenedProposals
+ */
+
+/**
+ * A map of the initialized wallets and their listened proposals.
+ * The keys are the wallet-ids
+ * @type {Record<string,WalletListenedProposals>}
+ */
+const walletListenedProposals = {};
 ```
 
-Should any request for this proposal return an unrecoverable error, for example a `404` response that indicates the proposal has already expired, it should be automatically removed from memory.
+For simplicity, references to `listenedProposals` on this document refer to the listened proposals for a specific initialized wallet
 
-### Create
+Should any request for a proposal return an unrecoverable error, for example a `404` response that indicates the proposal has already expired, it should be automatically removed from memory.
```

Comment:
How are we planning to do this?
Since only some routes will have the proposalId we could have a middleware to load the proposal from the `listenedProposals` and if an error is raised, e.g. `AtomicSwapError` or `ProposalUnrecoverableFaillure`, we can have the cleanup process on the middleware instead of on each method.

The downside of this is that some routes expect the proposal id as a body argument and others as a route parameter.
I think there are 2 alternatives:

1. We could have the `routes` function extract the proposalId and pass it to the middleware.
2. Move all routes that use proposalId to receive in the same way, e.g. `X-PROPOSAL-ID` header or on the querystring

The option 1. would be something along the lines of:

```javascript
router.get(
  '/register/:proposalId',
  registerSchema,
  async (req, res) => {
    return atomicSwapProposalWrapper(req.params.proposalId, registerController)(req, res);
  },
);
```

And the second can be a normal middleware, but at the cost of flexibility on the api

Reply:
After a more careful listing of all possible failures, I concluded the only irrecoverable one is the `404` from the backend on the `register`, `get` and `update` routes. 
Besides that, only on the `register`, the `403` error indicates the user has actually informed wrong inputs for the registration.

This is because the `register` endpoint already guarantees that any proposal inside it will contain the correct password, so no further validation is required at the other endpoints.

Because of that, I believe having a middleware would not be necessary, as the controllers themselves can have the necessary logic without becoming bloated.

Do you agree?